Fix the unsoundness in the `early_otherwise_branch` mir opt pass

JakobDegen opened the pull request:
Closes #78496 .

This change is a significant rewrite of much of the pass. Exactly what it does is documented in the source file (with ascii art!), and all the changes that are made to the MIR that are not trivially sound are carefully documented. That being said, this is my first time touching MIR, so there are probably some invariants I did not know about that I broke.

This version of the optimization is also somewhat more flexible than the original; for example, we do not care how or where the value on which the parent is switching is computed. There is no requirement that any types be the same. This could be made even more flexible in the future by allowing a wider range of statements in the bodies of `BBC, BBD` (as long as they are all the same of course). This should be a good first step though.

Probably needs a perf run.

r? @oli-obk who reviewed things the last time this was touched

JakobDegen commented on compiler/rustc_mir_transform/src/early_otherwise_branch.rs:
```diff
@@ -167,201 +224,170 @@ impl<'tcx> MirPass<'tcx> for EarlyOtherwiseBranch {
+            //  - In `BBC.2, BBD.2`: Not possible, since `discriminant(P)` was computed prior to
+            //    reaching that block in the input to our transformation, and so any data
+            //    invalidated by that computation could not have been used there.
+            //  - In `BB9`: Not possible since control flow might have reached `BB9` via the
+            //    `otherwise` branch in `BBC, BBD` in the input to our transformation, which would
+            //    have invalidated the data when computing `discriminant(P)`
```
I realize now that depending on the MIR semantics, this may be incorrect. In particular, we can imagine BB9 doing something like this:
```
if Q == otherwise {
    use reference invalidated by P
}
```
This would have been correct before this pass, since `P` is only computed if `Q` does not take the `otherwise` branch; it's obviously no longer correct after the pass. That being said, this kind of a pattern can't ever pass borrowck (since borrowck will not consider that the conditions in the branch in BB1 and BB9 are the same). Is this a legal transformation for other optimizations to make? Or must the validity of all references (whenever they're used) in the MIR always follow only from the control flow? (hopefully I explained myself accurately there)

oli-obk replied:
Great explanation, thanks. Unfortunately I don't have a good answer. MIR semantics are very underdocumented. As far as I know there is no opt that could cause such a pattern right now, but I'm wondering if we could write it with raw pointers and intrinsics

oli-obk commented:
Will review some time this week

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 396d8c60dbd6cb21d2176ac43f9f542c37fc404b with merge 236e8f85e050e93c72957ee20322d47010a24084...
<!-- homu: {"type":"TryBuildStarted","head_sha":"396d8c60dbd6cb21d2176ac43f9f542c37fc404b","merge_sha":"236e8f85e050e93c72957ee20322d47010a24084"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4503716465?check_suite_focus=true)
Build commit: 236e8f85e050e93c72957ee20322d47010a24084 (`236e8f85e050e93c72957ee20322d47010a24084`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4503716465?check_suite_focus=true"},"merge_sha":"236e8f85e050e93c72957ee20322d47010a24084"} -->

rust-timer commented:
Queued 236e8f85e050e93c72957ee20322d47010a24084 with parent f7fd79ac1d485ab47b62146f6dafed4aad5d9c6d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f7fd79ac1d485ab47b62146f6dafed4aad5d9c6d&end=236e8f85e050e93c72957ee20322d47010a24084).

rust-timer commented:
Finished benchmarking commit (236e8f85e050e93c72957ee20322d47010a24084): [comparison url](https://perf.rust-lang.org/compare.html?start=f7fd79ac1d485ab47b62146f6dafed4aad5d9c6d&end=236e8f85e050e93c72957ee20322d47010a24084).

**Summary**: This change led to very large relevant regressions 😿 in compiler performance.
- Very large regression in instruction counts (up to 6.3% on `incr-unchanged` builds of `inflate`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

simonvandel commented on compiler/rustc_mir_transform/src/early_otherwise_branch.rs:
```diff
@@ -15,144 +15,201 @@ use super::simplify::simplify_cfg;
     fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
-        //  FIXME(#78496)
-        sess.opts.debugging_opts.unsound_mir_opts && sess.mir_opt_level() >= 3
+        sess.mir_opt_level() >= 3
```
I'm pretty sure the opt level gate needs to be lowered to see useful perf results. Reducing it to 1 should do the trick.

JakobDegen commented:
This is causing issues with incremental and there are other things I want to look at too before this gets merged. I can close and re-open instead if that's preferable

JakobDegen commented:
I've dropped the MIR-opt-level requirement to 2 so that this actually runs. Also fixed an ICE and made a couple of other adjustments. This is still a draft because it's [causing issues with incremental](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/MIR.20opt.20setting.20off.20inconsistent.20fingerprint.20check.20failures/near/264841102), but besides that I think it should be ready for review.

JakobDegen commented:
Realize I forgot to comment on this: There had been a suggestion on the original version of this pass to do a `BitAnd`, thereby avoiding the extra branch and bb, but that's unfortunately not correct if `0` is one of the values that gets switched on. I could special case in a check for that, but I'm not sure the complexity is warranted.

oli-obk commented:
> I also did local perf runs of this and it was a flat zero across the board, so I'm not expecting this to bring many benefits. Having the opt be sound should be a good step forward regardless though.

Indeed, I just wanted to make sure we don't regress anything. I also agree with the "spurious noise" assessment on the one regression

JakobDegen commented:
@oli-obk will you get a chance to review this, or is there someone else I should ask? If there's no review capacity right now, I can close this until there is if that's preferable

oli-obk commented:
Sorry, didn't get to this properly before Christmas. I'll get back to reviewing complex things like this mid January

camelid commented:
triage: friendly ping @oli-obk :)

oli-obk commented:
Hmm I reviewed this before but obviously never wrote any conclusions. Well, third time's the charm. The opt is really well documented and structured and I feel confident it actually catches all the cases where it shouldn't run.

@bors r+ rollup=never

bors commented:
:pushpin: Commit c0c13b709e4a5ffb2ef1799d606719b4b5ad9e3a has been approved by `oli-obk`

<!-- @bors r=oli-obk c0c13b709e4a5ffb2ef1799d606719b4b5ad9e3a -->
<!-- homu: {"type":"Approved","sha":"c0c13b709e4a5ffb2ef1799d606719b4b5ad9e3a","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit c0c13b709e4a5ffb2ef1799d606719b4b5ad9e3a with merge a7f375789bab1a4e4a291c963081a8ca7d2b6bd7...
<!-- homu: {"type":"BuildStarted","head_sha":"c0c13b709e4a5ffb2ef1799d606719b4b5ad9e3a","merge_sha":"a7f375789bab1a4e4a291c963081a8ca7d2b6bd7"} -->

rust-timer commented:
Finished benchmarking commit (a7f375789bab1a4e4a291c963081a8ca7d2b6bd7): [comparison url](https://perf.rust-lang.org/compare.html?start=788b1fe5b79a8b74215022f9df49b0eae68a50b9&end=a7f375789bab1a4e4a291c963081a8ca7d2b6bd7).

**Summary**: This benchmark run shows 99 relevant improvements 🎉 to instruction counts.
- Average relevant regression: 1.3%
- Average relevant improvement: -1.1%
- Largest improvement in instruction counts: -6.9% on `full` builds of `deeply-nested-async check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
